```sklearn.decomposition``` 模块的 ```PCA``` 函数实现主成分分析，调用格式如下：

```sklearn.decomposition.PCA(n_components=None, copy=True)```（decomposition分解 PCA即Principal Components Analysis主成分分析）

其中，
n_components: 类型为 int 或字符串  

缺省时默认为 None，所有成分被保留  
赋值为 int，比如 n_components=2,提取两个主成分  
赋值为(0, 1)上的浮点数，自动选择主成分的个数，满足信息贡献率的要求（svd_solver='full）

copy：类型为 bool，即 True 或 False表示是否在运行算法时，将原始数据复制一份  

缺省时默认为True，表示运行PCA算法后，原始训练数据的值不会有任何改变，因为是在原始数据的副本上进行运算  

若为False，表示运行PCA算法后，原始训练数据的值会改变，因为是在原始数据上进行降维计算



In [20]:
import numpy as np
from sklearn.decomposition import PCA
# from sklearn.preprocessing import StandardScaler  # 标准化scaler比例尺，缩放

a = np.loadtxt("Pdata11_7.txt")
b = np.r_[a[:, 1:4], a[:, -3:]]  # 构造数据矩阵
scaler = StandardScaler()
b_scaled = scaler.fit_transform(b)

pca = PCA()  # 构造模型
md = pca.fit(b)  # 训练模型

print(f"特征值为: {md.explained_variance_}")
print(f"各主成分的贡献率: {md.explained_variance_ratio_}")
print(f"奇异值为: {md.singular_values_}")
print(f"各主成分的系数为:\n{md.components_}")  # 每行是一个主成分的系数
"""
下面直接计算特征值和特征向量, 和PCA对比
"""
covariance_matrix = np.cov(b.T)  # 协方差矩阵
eigenvalues, eigenvectors = np.linalg.eig(covariance_matrix)  # 求特征值和特征向量

print(f"特征值: {eigenvalues}")
print(f"特征向量:\n{eigenvectors}")  # 每一列的特征向量变成了主成分的一行的各个系数（行向量）
print(f"各主成分的贡献率为: {eigenvalues / np.sum(eigenvalues)}")

特征值为: [110.00413886  25.32447973   1.56804807]
各主成分的贡献率: [0.80355601 0.18498975 0.01145425]
奇异值为: [31.46485738 15.09703009  3.75665179]
各主成分的系数为:
[[ 0.55915657  0.42128705  0.71404562]
 [ 0.82767368 -0.33348264 -0.45138188]
 [ 0.04796048  0.84338992 -0.53515721]]
特征值: [110.00413886  25.32447973   1.56804807]
特征向量:
[[ 0.55915657  0.82767368 -0.04796048]
 [ 0.42128705 -0.33348264 -0.84338992]
 [ 0.71404562 -0.45138188  0.53515721]]
各主成分的贡献率为: [0.80355601 0.18498975 0.01145425]


注 11.1   
(1) 从上面程序运行结果可以看出， PCA 函数使用**协方差矩阵**作的主成分分析. 主成分分析也可以使用**相关系数矩阵**，两者计算结果略有差异，使用相关系数阵作主成分分析，相当千对数据进行了标准化处理．  

(2) 从程序的运行结果看，主成分的系数可以相差一个负号，因为特征向量乘以－1 后仍然为特征向量．用主成分分析做评价等其他模型时，觉得很难把握，主成分的系数变成了相反数，模型的解释都变了．  
各主成分的贡献率分别为 80.36%, 18.50%, 1.15%．因此， 前两个主成分的累积贡献率已达 98.86%， 应用中可取前两个主成分  

$F_1 = 0.5592x_1 + 0.4213x_2 + 0. 7140x_3$  

$F_2 = 0.8277x_1 - 0.3335x_2 - 0.4514x_3$ 

第一主成分 $F_1$ 是身高值 $x_1$、胸围值 $x_2$ 和体重值 $x_3$ 的加权和，当一个学生的 $F_1$值较大时， 可以推断他较高或较胖或又高又胖；反之，当一个学生比较魁梧时，所对应的尺值一般也较大， 故第一主成分是反映学生身材是否魁梧的综合指标，一般称之为 “大小” 因子. 而在第二主成分的表达式中， 身高$x_1$ 前的系数为正， 而胸围 $x_2$ 和体重 $x_3$ 前的系数为负， 当一个学生的$F_1$值较大时，说明 $x_1$ 的值较大， 而$x_2$ 和 $x_3$ 的值相对较小，即该生较高且瘦；反之， 瘦高型学生的 $F_2$ 值会较大， 故$F_2$是反映学生体型特征的综合指标，一般称之为＂形状”因子．

**例 11.8** 根据 2008 年安徽省统计年鉴资料，选择 $x_1$(工业总产值的现价)、$x_2$(工业销售按当年价的产值)、$x_3$(流动资产年平均余额)、$x_4$(固定资产净值年平均余额)、$x_5$(业务收入)、$x_6$(利润总额) 6 项指标进行主成分分析，表 11.4 列出了安徽省各市大中型工业企业主要经济指标的统计数据. 
(1) 选取指标是否合适？
(2) 给出各市大中型工业企业排名. 

表 11.4 安徽省各市大中型工业企业主要经济指标（单位：亿元）

| 地区 | $x_1$ | $x_2$ | $x_3$ | $x_4$ | $x_5$ | $x_6$ |
| ---- | ---- | ---- | ---- | ---- | ---- | ---- |
| 合肥市 | 1932.27 | 1900.53 | 653.83 | 570.95 | 1810.70 | 119.53 |
| 淮北市 | 367.05 | 366.08 | 186.16 | 252.07 | 395.43 | 32.82 |
| 亳州市 | 86.89 | 85.38 | 40.85 | 51.71 | 83.26 | 8.95 |
| 宿州市 | 154.27 | 147.07 | 30.68 | 57.96 | 146.30 | $-1.27$ |
| 蚌埠市 | 197.21 | 193.28 | 104.56 | 90.15 | 182.60 | 7.85 |
| 阜阳市 | 244.17 | 231.55 | 56.37 | 121.96 | 224.04 | 26.49 |
| 淮南市 | 497.74 | 483.69 | 206.80 | 501.37 | 496.59 | 27.76 |
| 滁州市 | 308.91 | 296.99 | 118.65 | 76.90 | 277.42 | 19.32 |
| 六安市 | 191.77 | 189.05 | 70.19 | 62.31 | 191.98 | 23.08 |
| 马鞍山市 | 905.32 | 894.61 | 351.52 | 502.99 | 1048.02 | 53.88 |
| 巢湖市 | 254.99 | 242.38 | 106.66 | 75.48 | 234.76 | 19.65 |
| 芜湖市 | 867.07 | 852.34 | 418.82 | 217.76 | 806.94 | 37.01 |
| 宣城市 | 219.36 | 207.07 | 82.58 | 54.74 | 192.74 | 11.02 |
| 铜陵市 | 570.33 | 563.33 | 224.23 | 190.77 | 697.91 | 20.61 |
| 池州市 | 59.11 | 57.32 | 16.97 | 40.33 | 56.56 | 6.03 |
| 安庆市 | 430.58 | 426.25 | 103.08 | 147.05 | 442.04 | 0.79 |
| 黄山市 | 65.03 | 64.36 | 28.38 | 8.58 | 60.48 | 2.88 |

注：数据来源《安徽统计年鉴 2008》. 

解 (1) 利用 Python 软件，求得相关系数矩阵 

$$
R = \begin{bmatrix}
1.0000 & 1.0000 & 0.9754 & 0.8231 & 0.9914 & 0.9375 \\
1.0000 & 1.0000 & 0.9758 & 0.8236 & 0.9920 & 0.9369 \\
0.9754 & 0.9758 & 1.0000 & 0.8245 & 0.9712 & 0.9127 \\
0.8231 & 0.8236 & 0.8245 & 1.0000 & 0.8502 & 0.8020 \\
0.9914 & 0.9920 & 0.9712 & 0.8502 & 1.0000 & 0.9212 \\
0.9375 & 0.9369 & 0.9127 & 0.8020 & 0.9212 & 1.0000 \\
\end{bmatrix}
$$

由于 $r_{12} = r_{21} = 1$，表明指标 $x_1, x_2$ 完全线性相关，所以选取的指标不合适，只保留 $x_1, x_2$ 中的一个指标，这里删除指标 $x_1$. 

(2) 各市大中型工业企业排名的数学原理在这里就不赘述了. 我们只给出计算结果和 Python 程序. 

第一主成分信息贡献率达到 92.2%，选取一个主成分进行评价即可，主成分信息贡献率计算结果见表 11.5. 根据第一主成分的评价结果见表 11.6. 

表 11.5 特征值、特征向量及贡献率

| 特征值 | 特征向量 | 贡献率/% |
| ---- | ---- | ---- |
| 4.6100 | $[0.4595, 0.4552, 0.4158, 0.4600, 0.4441]^T$ | 92.2009 |
| 0.2475 | $[0.2517, 0.2103, -0.9054, 0.1315, 0.2354]^T$ | 4.9501 |
| 0.1050 | $[0.1926, 0.3702, -0.0390, 0.3029, -0.8559]^T$ | 2.1007 |
| 0.0322 | $[-0.3510, 0.7779, 0.0275, -0.5153, 0.0738]^T$ | 0.6431 |
| 0.0053 | $[-0.7518, 0.0803, -0.0719, 0.6434, 0.0965]^T$ | 0.1053 |

表 11.6 各市第一主成分得分排名

| 排名 | 地区 | 得分 | 排名 | 地区 | 得分 | 排名 | 地区 | 得分 |
| ---- | ---- | ---- | ---- | ---- | ---- | ---- | ---- | ---- |
| 1 | 合肥市 | 6.2827 | 7 | 安庆市 | $-0.5654$ | 13 | 宣城市 | $-1.1219$ |
| 2 | 马鞍山市 | 2.6810 | 8 | 滁州市 | $-0.6892$ | 14 | 亳州市 | $-1.4888$ |
| 3 | 芜湖市 | 1.6979 | 9 | 阜阳市 | $-0.7568$ | 15 | 宿州市 | $-1.5324$ |
| 4 | 淮南市 | 0.9914 | 10 | 巢湖市 | $-0.8118$ | 16 | 池州市 | $-1.6711$ |
| 5 | 铜陵市 | 0.5144 | 11 | 六安市 | $-0.9758$ | 17 | 黄山市 | $-1.7484$ |
| 6 | 淮北市 | 0.2516 | 12 | 蚌埠市 | $-1.0575$ |  |  |  | 

把表 11.4 中的  的 6 列 17 行数据保存到文本文件 Pdata11_8.txt 中 

In [ ]:
import numpy as np
from scipy.stats import zscore

a = np.loadtxt("Pdata11_8.txt")
b = np.delete(a, 0, axis=1)  # 删除第一列数据
c = zscore(b)  # z-score也叫标准分数，用来标准化
r = np.corrcoef(c.T)  # 计算相关系数矩阵，correlation coeffitient相关系数corrcorf
d, e = np.linalg.eig(r)  # 计算特征值和特征向量
rate = d / d.sum()  # 计算各主成分的贡献率

print(f"特征值为: {d}")
print(f"特征向量为:\n{e}")
print(f"各主成分的贡献率为: {rate}")

k = 1  # 提出主成分的个数，根据各主成分的得分率取得
F = e[:, :k]  # 第1列（索引为0）所有行，即第一列全部元素，特征向量u_1
score_mat = c.dot(F)  # 计算主成分得分矩阵，向量点积
score1 = score_mat.dot(rate[0:k])  # 计算各评价对象的得分,rate[0:k]=0.92200854
score2 = - score1  # 通过观测，调整得分的正负号

print(f"各评价对象的的得分为:\n{score2}")

# array.argsort()返回排序后各元素在原数组的原始索引（类似秩统计量）+1 确保索引从1开始
index = score1.argsort() + 1  # 排序后的每个元素在原数组中的位置

print(f"从高到低各个城市的编号顺序为: {index}")


特征值为: [4.61004272 0.24750271 0.10503409 0.00526533 0.03215514]
特征向量为:
[[-0.45952752 -0.25173654 -0.1925837   0.75177158  0.35101879]
 [-0.45518408 -0.21034324 -0.37016683 -0.08028926 -0.77787745]
 [-0.41576952  0.90537078  0.03899078  0.07185465 -0.02749689]
 [-0.45997279 -0.13147819 -0.30288681 -0.64336245  0.51525028]
 [-0.44407506 -0.23537559  0.85593653 -0.09651316 -0.07384811]]
各主成分的贡献率为: [0.92200854 0.04950054 0.02100682 0.00105307 0.00643103]
各评价对象的的得分为:
[ 6.28265085  0.2516401  -1.4888003  -1.53237715 -1.05752141 -0.7568012
  0.99141051 -0.68917209 -0.97584699  2.68099041 -0.8118024   1.69791715
 -1.12187592  0.51443663 -1.67112384 -0.56535826 -1.74836609]
从高到低各个城市的编号顺序为: [ 1 10 12  7 14  2 16  8  6 11  9  5 13  3  4 15 17]
